### PR TITLE
Refactor tests for mocha

### LINKED ISSUE #137: Replace the vows test framework as it is no longer being supported
The vows testing framework being used has not been updated for a long time now.
We need to move on and use a different framework.

Possible candidates would be

- mocha
- jest

~~~While I like mocha, jest seems to be the better option as it uses child processes for running individual test cases. With jest in place, we could get rid of the process spawning facility now in place and simplify existing test cases. jest also exploits #129. Besides that, I assume that the usage of the jest testing framework will grow over time.~~~

If for myself would recommend mocha as we do not require the additional features (react testing) offered by jest.

@raszi what do you think?

### PR DESCRIPTION
**Pull request by silkentrance**
fix #137

WIP

- [x] cleanups for outband tests are still missing
- [x] failing discardDescriptor tests require #149
- [x] drop support for node < 0.10.0

### CODE REVIEW

**Comment by silkentrance** [comment]
Tests against the template option (here:effective gid mismatch) failed partly due to the fact that tmp will not join paths with opts.dir nor the default tmpDir. This resulted in the fact that  the temporary files and directories will be created in the CWD.

I will file a new issue for this.

**Comment by raszi** [comment]
Thank you @silkentrance for the hard work, this will help!

**Comment by silkentrance** [comment]
@raszi You're welcome. The existing tests that spawn child processes are still missing. I need to go through the code to find out how to do it with mocha...

**Comment by silkentrance** [comment]
rebased to master - failing tests are due to the fact that when passing a template that has no leading directory component in the path, we should consider creating the tmp file in the default tmpdir instead of the CWD.

What do you think?

**Comment by silkentrance** [comment]
@raszi I will look over the existing issues with the code and commit these then as I already merged into master.

File: test/assertions.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+var
+  assert = require('assert'),
+  fs = require('fs'),
+  path = require('path'),
+  existsSync = fs.existsSync || path.existsSync;
+
+
+module.exports.assertName = function assertName(name, expected) {
+  assert.ok(typeof name == 'string');
+  assert.ok(name.length > 0, 'an empty string is not a valid name');
```
[the remark this answers is not in the document]
Comment (silkentrance) [reply test/assertions.js]: done.

File: test/child-process.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  fs = require('fs'),
+  path = require('path'),
+  existsSync = fs.existsSync || path.existsSync,
+  spawn = require('child_process').spawn;
+
+
+module.exports = function spawnChildProcess(configFile, cb) {
+  var
+    node_path = process.argv[0],
+    command_args = [ path.join(__dirname, 'spawn.js') ].concat(configFile),
```
[the remark this answers is not in the document]
Comment (silkentrance) [reply test/child-process.js]: done.

File: test/dir-sync-test.js [hunk 3]
```diff
@@ -1,232 +1,139 @@
...
+    describe('when running inband standard tests', function () {
+
+      inbandStandardTests(false, function before() {
+        this.topic = tmp.dirSync(this.opts);
+      });
+
+      describe('with invalid tries', function () {
+        it('should result in an error on negative tries', function () {
+          try {
+            tmp.dirSync({tries: -1});
```
[the remark this answers is not in the document]
Comment (silkentrance) [reply test/dir-sync-test.js]: @raszi like so?

```
        it('should result in an error on negative tries', function () {

          try {
            tmp.dirSync({tries: -1});
            assert.fail('should have failed');
          } catch (err) {
            assert.ok(err instanceof Error);
          }

        });
```

**Pull request merged**